File: SWEagent/tests/test-quick-stats.test.ts
/**
 * Quick stats tests converted from test_quick_stats.py
 */

import { describe, it, expect, beforeEach, afterEach } from '@jest/globals';
import { quickStats } from '../src/run/quick-stats';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

describe('Quick Stats', () => {
  let tmpDir: string;

  beforeEach(() => {
    // Create a temporary directory for test files
    tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'quick-stats-test-'));
  });

  afterEach(() => {
    // Clean up temporary directory
    if (fs.existsSync(tmpDir)) {
      fs.rmSync(tmpDir, { recursive: true });
    }
  });

  describe('quickStats function', () => {
    it('should handle empty directories', async () => {
      const result = await quickStats(tmpDir);
      expect(result).toBe('No .traj files found.');
    });

    it('should process trajectory files correctly', async () => {
      // Create a sample .traj file with required structure
      const trajFile = path.join(tmpDir, 'test.traj');
      const trajData = {
        info: {
          model_stats: {
            api_calls: 42,
            tokens_sent: 1000,
            tokens_received: 500,
            instance_cost: 0.05,
          },
          exit_status: 'success',
        },
      };

      fs.writeFileSync(trajFile, JSON.stringify(trajData));

      const result = await quickStats(tmpDir);

      // Check that the result contains our exit status
      expect(result).toContain('## `success`');
      expect(result).not.toBe('No .traj files found.');
    });

    it('should handle multiple trajectory files', async () => {
      // Create multiple .traj files with different exit statuses
      const trajData1 = {
        info: {
          model_stats: {
            api_calls: 10,
            tokens_sent: 500,
            tokens_received: 200,
            instance_cost: 0.02,
          },
          exit_status: 'success',
        },
      };

      const trajData2 = {
        info: {
          model_stats: {
            api_calls: 5,
            tokens_sent: 300,
            tokens_received: 150,
            instance_cost: 0.01,
          },
          exit_status: 'exit_cost',
        },
      };

      const trajData3 = {
        info: {
          model_stats: {
            api_calls: 20,
            tokens_sent: 1500,
            tokens_received: 800,
            instance_cost: 0.08,
          },
          exit_status: 'success',
        },
      };

      fs.writeFileSync(path.join(tmpDir, 'test1.traj'), JSON.stringify(trajData1));
      fs.writeFileSync(path.join(tmpDir, 'test2.traj'), JSON.stringify(trajData2));
      fs.writeFileSync(path.join(tmpDir, 'test3.traj'), JSON.stringify(trajData3));

      const result = await quickStats(tmpDir);

      // Check that both exit statuses are present
      expect(result).toContain('## `success`');
      expect(result).toContain('## `exit_cost`');

      // Check that counts are correct
      expect(result).toMatch(/`success`.*2 trajectories/);
      expect(result).toMatch(/`exit_cost`.*1 trajectories/);
    });

    it('should handle malformed trajectory files gracefully', async () => {
      // Create a malformed .traj file
      const malformedFile = path.join(tmpDir, 'malformed.traj');
      fs.writeFileSync(malformedFile, 'not valid json');

      // Create a valid .traj file
      const validFile = path.join(tmpDir, 'valid.traj');
      const validData = {
        info: {
          model_stats: {
            api_calls: 5,
          },
          exit_status: 'success',
        },
      };
      fs.writeFileSync(validFile, JSON.stringify(validData));

      const result = await quickStats(tmpDir);

      // Should still process the valid file
      expect(result).toContain('## `success`');
    });

    it('should handle trajectory files without model_stats', async () => {
      const trajData = {
        info: {
          exit_status: 'submitted',
        },
      };

      fs.writeFileSync(path.join(tmpDir, 'test.traj'), JSON.stringify(trajData));

      const result = await quickStats(tmpDir);

      expect(result).toContain('## `submitted`');
    });

    it('should handle trajectory files without exit_status', async () => {
      const trajData = {
        info: {
          model_stats: {
            api_calls: 10,
          },
        },
      };

      fs.writeFileSync(path.join(tmpDir, 'test.traj'), JSON.stringify(trajData));

      const result = await quickStats(tmpDir);

      // Should handle gracefully, possibly with 'unknown' or similar status
      expect(result).not.toBe('No .traj files found.');
    });

    it('should calculate statistics correctly', async () => {
      // Create files with known stats for verification
      const trajData1 = {
        info: {
          model_stats: {
            api_calls: 10,
            tokens_sent: 1000,
            tokens_received: 500,
            instance_cost: 0.05,
          },
          exit_status: 'success',
        },
      };

      const trajData2 = {
        info: {
          model_stats: {
            api_calls: 20,
            tokens_sent: 2000,
            tokens_received: 1000,
            instance_cost: 0.1,
          },
          exit_status: 'success',
        },
      };

      fs.writeFileSync(path.join(tmpDir, 'test1.traj'), JSON.stringify(trajData1));
      fs.writeFileSync(path.join(tmpDir, 'test2.traj'), JSON.stringify(trajData2));

      const result = await quickStats(tmpDir);

      // Check that stats are aggregated correctly
      expect(result).toContain('success');

      // The exact format might vary, but we should see totals or averages
      // This depends on the implementation of quickStats
    });

    it('should handle nested directories', async () => {
      // Create a nested directory structure
      const nestedDir = path.join(tmpDir, 'nested', 'deep');
      fs.mkdirSync(nestedDir, { recursive: true });

      const trajData = {
        info: {
          model_stats: {
            api_calls: 5,
          },
          exit_status: 'success',
        },
      };

      // Put trajectory files at different levels
      fs.writeFileSync(path.join(tmpDir, 'root.traj'), JSON.stringify(trajData));
      fs.writeFileSync(path.join(nestedDir, 'nested.traj'), JSON.stringify(trajData));

      const result = await quickStats(tmpDir);

      // Should find both files
      expect(result).toContain('## `success`');
      expect(result).toMatch(/2 trajectories/);
    });

    it('should sort exit statuses consistently', async () => {
      // Create files with various exit statuses
      const statuses = ['exit_cost', 'success', 'submitted', 'exit_format', 'exit_context'];

      statuses.forEach((status, index) => {
        const trajData = {
          info: {
            model_stats: {
              api_calls: index + 1,
            },
            exit_status: status,
          },
        };
        fs.writeFileSync(path.join(tmpDir, `test${index}.traj`), JSON.stringify(trajData));
      });

      const result = await quickStats(tmpDir);

      // Check that all statuses are present
      statuses.forEach((status) => {
        expect(result).toContain(`## \`${status}\``);
      });

      // Check ordering (should be alphabetical or by frequency)
      const statusOrder = result.match(/## `([^`]+)`/g);
      expect(statusOrder).toBeDefined();
      expect(statusOrder!.length).toBe(statuses.length);
    });

    it('should handle special characters in exit status', async () => {
      const trajData = {
        info: {
          model_stats: {
            api_calls: 5,
          },
          exit_status: 'submitted (exit_cost)',
        },
      };

      fs.writeFileSync(path.join(tmpDir, 'test.traj'), JSON.stringify(trajData));

      const result = await quickStats(tmpDir);

      expect(result).toContain('submitted (exit_cost)');
    });

    it('should provide meaningful output format', async () => {
      // Create a few trajectory files
      const trajData = {
        info: {
          model_stats: {
            api_calls: 10,
            tokens_sent: 1000,
            tokens_received: 500,
            instance_cost: 0.05,
          },
          exit_status: 'success',
        },
      };

      fs.writeFileSync(path.join(tmpDir, 'test1.traj'), JSON.stringify(trajData));
      fs.writeFileSync(path.join(tmpDir, 'test2.traj'), JSON.stringify(trajData));

      const result = await quickStats(tmpDir);

      // Should have markdown formatting
      expect(result).toMatch(/##/); // Markdown headers
      expect(result).toContain('trajectories'); // Trajectory count

      // Should potentially show averages or totals
      // The exact format depends on implementation
    });
  });
});

File: cpp/generated_testing/src/test-quick-stats.test.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/SWEagent/tests/test-quick-stats.test.h"

void Main(void)
{
    describe(std::string("Quick Stats"), [=]() mutable
    {
        shared<string> tmpDir;
        beforeEach([=]() mutable
        {
            tmpDir = fs::mkdtempSync(path->join(os::tmpdir(), std::string("quick-stats-test-")));
        }
        );
        afterEach([=]() mutable
        {
            if (fs::existsSync(tmpDir)) {
                fs::rmSync(tmpDir, object{
                    object::pair{std::string("recursive"), true}
                });
            }
        }
        );
        describe(std::string("quickStats function"), [=]() mutable
        {
            it(std::string("should handle empty directories"), [=]() mutable
            {
                auto result = std::async([=]() { quickStats(tmpDir); });
                expect(result)->toBe(std::string("No .traj files found."));
            }
            );
            it(std::string("should process trajectory files correctly"), [=]() mutable
            {
                auto trajFile = path->join(tmpDir, std::string("test.traj"));
                auto trajData = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 42}, 
                            object::pair{std::string("tokens_sent"), 1000}, 
                            object::pair{std::string("tokens_received"), 500}, 
                            object::pair{std::string("instance_cost"), 0.05}
                        }}, 
                        object::pair{std::string("exit_status"), std::string("success")}
                    }}
                };
                fs::writeFileSync(trajFile, JSON->stringify(trajData));
                auto result = std::async([=]() { quickStats(tmpDir); });
                expect(result)->toContain(std::string("## `success`"));
                expect(result)->not->toBe(std::string("No .traj files found."));
            }
            );
            it(std::string("should handle multiple trajectory files"), [=]() mutable
            {
                auto trajData1 = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 10}, 
                            object::pair{std::string("tokens_sent"), 500}, 
                            object::pair{std::string("tokens_received"), 200}, 
                            object::pair{std::string("instance_cost"), 0.02}
                        }}, 
                        object::pair{std::string("exit_status"), std::string("success")}
                    }}
                };
                auto trajData2 = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 5}, 
                            object::pair{std::string("tokens_sent"), 300}, 
                            object::pair{std::string("tokens_received"), 150}, 
                            object::pair{std::string("instance_cost"), 0.01}
                        }}, 
                        object::pair{std::string("exit_status"), std::string("exit_cost")}
                    }}
                };
                auto trajData3 = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 20}, 
                            object::pair{std::string("tokens_sent"), 1500}, 
                            object::pair{std::string("tokens_received"), 800}, 
                            object::pair{std::string("instance_cost"), 0.08}
                        }}, 
                        object::pair{std::string("exit_status"), std::string("success")}
                    }}
                };
                fs::writeFileSync(path->join(tmpDir, std::string("test1.traj")), JSON->stringify(trajData1));
                fs::writeFileSync(path->join(tmpDir, std::string("test2.traj")), JSON->stringify(trajData2));
                fs::writeFileSync(path->join(tmpDir, std::string("test3.traj")), JSON->stringify(trajData3));
                auto result = std::async([=]() { quickStats(tmpDir); });
                expect(result)->toContain(std::string("## `success`"));
                expect(result)->toContain(std::string("## `exit_cost`"));
                expect(result)->toMatch((new RegExp(std::string("`success`.*2 trajectorie"))));
                expect(result)->toMatch((new RegExp(std::string("`exit_cost`.*1 trajectorie"))));
            }
            );
            it(std::string("should handle malformed trajectory files gracefully"), [=]() mutable
            {
                auto malformedFile = path->join(tmpDir, std::string("malformed.traj"));
                fs::writeFileSync(malformedFile, std::string("not valid json"));
                auto validFile = path->join(tmpDir, std::string("valid.traj"));
                auto validData = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 5}
                        }}, 
                        object::pair{std::string("exit_status"), std::string("success")}
                    }}
                };
                fs::writeFileSync(validFile, JSON->stringify(validData));
                auto result = std::async([=]() { quickStats(tmpDir); });
                expect(result)->toContain(std::string("## `success`"));
            }
            );
            it(std::string("should handle trajectory files without model_stats"), [=]() mutable
            {
                auto trajData = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("exit_status"), std::string("submitted")}
                    }}
                };
                fs::writeFileSync(path->join(tmpDir, std::string("test.traj")), JSON->stringify(trajData));
                auto result = std::async([=]() { quickStats(tmpDir); });
                expect(result)->toContain(std::string("## `submitted`"));
            }
            );
            it(std::string("should handle trajectory files without exit_status"), [=]() mutable
            {
                auto trajData = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 10}
                        }}
                    }}
                };
                fs::writeFileSync(path->join(tmpDir, std::string("test.traj")), JSON->stringify(trajData));
                auto result = std::async([=]() { quickStats(tmpDir); });
                expect(result)->not->toBe(std::string("No .traj files found."));
            }
            );
            it(std::string("should calculate statistics correctly"), [=]() mutable
            {
                auto trajData1 = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 10}, 
                            object::pair{std::string("tokens_sent"), 1000}, 
                            object::pair{std::string("tokens_received"), 500}, 
                            object::pair{std::string("instance_cost"), 0.05}
                        }}, 
                        object::pair{std::string("exit_status"), std::string("success")}
                    }}
                };
                auto trajData2 = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 20}, 
                            object::pair{std::string("tokens_sent"), 2000}, 
                            object::pair{std::string("tokens_received"), 1000}, 
                            object::pair{std::string("instance_cost"), 0.1}
                        }}, 
                        object::pair{std::string("exit_status"), std::string("success")}
                    }}
                };
                fs::writeFileSync(path->join(tmpDir, std::string("test1.traj")), JSON->stringify(trajData1));
                fs::writeFileSync(path->join(tmpDir, std::string("test2.traj")), JSON->stringify(trajData2));
                auto result = std::async([=]() { quickStats(tmpDir); });
                expect(result)->toContain(std::string("success"));
            }
            );
            it(std::string("should handle nested directories"), [=]() mutable
            {
                auto nestedDir = path->join(tmpDir, std::string("nested"), std::string("deep"));
                fs::mkdirSync(nestedDir, object{
                    object::pair{std::string("recursive"), true}
                });
                auto trajData = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 5}
                        }}, 
                        object::pair{std::string("exit_status"), std::string("success")}
                    }}
                };
                fs::writeFileSync(path->join(tmpDir, std::string("root.traj")), JSON->stringify(trajData));
                fs::writeFileSync(path->join(nestedDir, std::string("nested.traj")), JSON->stringify(trajData));
                auto result = std::async([=]() { quickStats(tmpDir); });
                expect(result)->toContain(std::string("## `success`"));
                expect(result)->toMatch((new RegExp(std::string("2 trajectorie"))));
            }
            );
            it(std::string("should sort exit statuses consistently"), [=]() mutable
            {
                auto statuses = array<string>{ std::string("exit_cost"), std::string("success"), std::string("submitted"), std::string("exit_format"), std::string("exit_context") };
                statuses->forEach([=](auto status, auto index) mutable
                {
                    auto trajData = object{
                        object::pair{std::string("info"), object{
                            object::pair{std::string("model_stats"), object{
                                object::pair{std::string("api_calls"), index + 1}
                            }}, 
                            object::pair{std::string("exit_status"), status}
                        }}
                    };
                    fs::writeFileSync(path->join(tmpDir, std::string("test") + index + std::string(".traj")), JSON->stringify(trajData));
                }
                );
                shared result = std::async([=]() { quickStats(tmpDir); });
                statuses->forEach([=](auto status) mutable
                {
                    expect(result)->toContain(std::string("## `") + status + std::string("`"));
                }
                );
                auto statusOrder = result->match((new RegExp(std::string("## `([^`]+)`"))));
                expect(statusOrder)->toBeDefined();
                expect(statusOrder->length)->toBe(statuses->get_length());
            }
            );
            it(std::string("should handle special characters in exit status"), [=]() mutable
            {
                auto trajData = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 5}
                        }}, 
                        object::pair{std::string("exit_status"), std::string("submitted (exit_cost)")}
                    }}
                };
                fs::writeFileSync(path->join(tmpDir, std::string("test.traj")), JSON->stringify(trajData));
                auto result = std::async([=]() { quickStats(tmpDir); });
                expect(result)->toContain(std::string("submitted (exit_cost)"));
            }
            );
            it(std::string("should provide meaningful output format"), [=]() mutable
            {
                auto trajData = object{
                    object::pair{std::string("info"), object{
                        object::pair{std::string("model_stats"), object{
                            object::pair{std::string("api_calls"), 10}, 
                            object::pair{std::string("tokens_sent"), 1000}, 
                            object::pair{std::string("tokens_received"), 500}, 
                            object::pair{std::string("instance_cost"), 0.05}
                        }}, 
                        object::pair{std::string("exit_status"), std::string("success")}
                    }}
                };
                fs::writeFileSync(path->join(tmpDir, std::string("test1.traj")), JSON->stringify(trajData));
                fs::writeFileSync(path->join(tmpDir, std::string("test2.traj")), JSON->stringify(trajData));
                auto result = std::async([=]() { quickStats(tmpDir); });
                expect(result)->toMatch((new RegExp(std::string("#"))));
                expect(result)->toContain(std::string("trajectories"));
            }
            );
        }
        );
    }
    );
}

MAIN
